src/cargo/core/resolver/mod.rs

```diff
@@ -825,7 +825,9 @@ fn activation_error(
     conflicting_activations: &HashMap<PackageId, ConflictReason>,
     candidates: &[Candidate],
     config: Option<&Config>,
-) -> CargoError {
+) -> PackageError {
+    let to_package_err = |err| PackageError::new(err, parent.package_id().clone());
```

Comment:
I am not entirely sure what you are trying to do hear, but I doubt that the `parent` is the `package_id` that you want. If the problem occurs deep in the tree why do you need to know what the parent of the dependency that could not be resolved was? I would have thought you'd be looking for something in the workspace. But maybe I just don't understand what you are trying to do.

Reply:
In general my motive is the translation of cargo errors into IDE diagnostics in RLS. Right now the best I can do is generate a diagnostic for all cargo errors covering the root manifest. But if you mistype a dependency name in some workspace member's manifest, having a diagnostic on the root is hardly ideal.

The "parent" package id exposed allow me to infer that it's better to generate a diagnostic on the member's manifest than the root's. Similarly to how the error message does just that for humans.

ie:
```
error: no matching package named `nosuch` found                                                                                           
location searched: registry `https://github.com/rust-lang/crates.io-index`
required by package `rusttmp-member v0.1.0 (/home/alex/project/rusttmp/member)`
```

Also see https://github.com/rust-lang/cargo/issues/6144, https://github.com/rust-lang-nursery/rls/issues/1069

Reply:
perhaps the method being `parent_package_id()` would resolve the confusion in intention?

Reply:
Except the parent is not necessarily in the workspace. To take an example from the real world:
a toml like: 
```toml
[dependencies]
piston_window = "0.35.0"
piston2d-graphics = "0.24.0"
pistoncore-glutin_window = "0.43.0"
piston2d-opengl_graphics = "0.50.0"
```

leads to the error:

```cmd
error: failed to select a version for `x11-dl`.
    ... required by package `glutin v0.4.2`
    ... which is depended on by `pistoncore-glutin_window v0.20.0`
    ... which is depended on by `piston_window v0.35.0`
    ... which is depended on by `cargo_speed_test v0.1.0 (C:\cargo_speed_test\test1)`
versions that meet the requirements `~2.0` are: 2.0.1, 2.0.0

all possible versions conflict with previously selected packages.

  previously selected package `x11-dl v2.18.3`
    ... which is depended on by `glutin v0.10.0`
    ... which is depended on by `pistoncore-glutin_window v0.43.0`
    ... which is depended on by `cargo_speed_test v0.1.0 (C:\cargo_speed_test\test1)`

failed to select a version for `x11-dl` which could resolve this conflict
```

The `parent` in is `glutin v0.4.2`. How is that useful to you? ( I will go away with a screenshot of your integrated RLS doing something reasonable on this real world case, in the case that I am to dance to see it. ) I'd think you would need `cargo_speed_test v0.1.0` (the library that is building this) or `piston_window v0.35.0` (something related to a row in the cargo.toml)

Reply:
Yes you're correct I can't always make use of it, in rls I only use this to direct the diagnostic if the parent package id is equal to some member's. Other cases will, for the moment, fallback to highlighting on the root.

I'd prefer to have a chain of errors from the root manifest as i have with `ManifestError`s, but I couldn't see how to easily add this here.

Reply:
Right now I've woken up, I've added the package id path-to-top to the `ResolveError`, instead of just the parent, which was pretty easy to do in the end and allows handling this case. Thank you very much for raising this example!

![](https://user-images.githubusercontent.com/2331607/47001279-fdfad400-d121-11e8-8f60-bf502cafbc63.jpg)

Reply:
Glad my stubbornness was helpful.